optuna/samplers/_search_space/intersection.py

```diff
@@ -59,20 +59,24 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
             if self._study_id != study._study_id:
                 raise ValueError("`IntersectionSearchSpace` cannot handle multiple studies.")
 
-        states_of_interest = [optuna.trial.TrialState.COMPLETE]
+        states_of_interest = [
+            optuna.trial.TrialState.COMPLETE,
+            optuna.trial.TrialState.WAITING,
+            optuna.trial.TrialState.RUNNING,
+        ]
 
         if self._include_pruned:
             states_of_interest.append(optuna.trial.TrialState.PRUNED)
 
-        next_cursor = self._cursor
-        for trial in reversed(study.get_trials(deepcopy=False)):
+        trials = study.get_trials(deepcopy=False, states=states_of_interest)
+
+        next_cursor = trials[-1].number + 1 if len(trials) > 0 else -1
```

Comment:
[question] It relies on that the assumption that `get_trials` returns a list of ordered trials, but is it actually guaranteed in all the storages? e.g. [redis](https://github.com/optuna/optuna/blob/master/optuna/storages/_redis.py#L687-L699), [SQL](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py#L953-L962)

Reply:
They are always wrapped by cached storage and trials are guaranteed to be ordered.
https://github.com/optuna/optuna/blob/master/optuna/storages/_cached_storage.py#L344

Reply:
Yes, we assume that `get_trials` do return the trials in chronological order. However, there is a known issue (#3605) that the returned trials can be unsorted. This issue is to be fixed.